packages/flutter/lib/src/rendering/layer.dart

```diff
@@ -135,6 +135,90 @@ class AnnotationResult<T> {
+  /// observers. For example, it may evaluate to true on a parent node but false
+  /// on a child if the parent has observers but the child does not.
+  ///
+  /// See also:
+  ///
+  ///   * [Layer.addCompositionCallback].
+  bool get subtreeHasCompositionCallbacks => _compositionCallbackCount > 0;
+
+  int _compositionCallbackCount = 0;
+  void _updateSubtreeCompositionObserverCount(int delta) {
+    assert(delta != 0);
+    _compositionCallbackCount += delta;
+    assert(_compositionCallbackCount >= 0);
+    if (parent != null) {
+      parent!._updateSubtreeCompositionObserverCount(delta);
+    }
+  }
+
+  void _fireCompositionCallbacks({required bool includeChildren}) {
+    for (final VoidCallback callback in List<VoidCallback>.of(_callbacks.values)) {
+      callback();
+    }
+  }
+
+  bool _debugMutationsLocked = false;
+
+  /// Describes the clip that would be applied to contents of this layer,
+  /// if any.
+  Rect? describeClipBounds() => null;
+
+  /// Adds a callback for when the layer tree that this layer is part of gets
+  /// composited, or when it is detached and will not be rendered again.
+  ///
+  /// This callback will fire even if an ancestor layer is added with retained
+  /// rendering, meaning that it will fire even if this layer gets added to the
+  /// scene via some call to [ui.SceneBuilder.addRetained] on one of its
+  /// ancestor layers.
+  ///
+  /// The callback receives a reference to this layer. The recipient must not
+  /// mutate the layer during the scope of the callback, but may traverse the
+  /// tree to find information about the current transform or clip. The layer
+  /// may not be [attached] anymore in this state, but even if it is detached it
+  /// may still have an also detached parent it can visit.
+  ///
+  /// If new callbacks are added or removed within the [callback], the new
+  /// callbacks will fire (or stop firing) on the _next_ compositing event.
+  ///
+  /// {@template flutter.rendering.Layer.compositionCallbacks}
+  /// Composition callbacks are useful in place of pushing a layer that would
+  /// otherwise try to observe the layer tree without actually affecting
+  /// compositing. For example, a composition callback may be used to observe
+  /// the total transform and clip of the current container layer to determine
+  /// whether a render object drawn into it is visible or not.
+  ///
+  /// Calling the returned callback will remove [callback] from the composition
+  /// callbacks.
+  /// {@endtemplate}
+  VoidCallback addCompositionCallback(CompositionCallback callback) {
+    _updateSubtreeCompositionObserverCount(1);
+    final int callbackId = _nextCallbackId += 1;
+    _callbacks[callbackId] = () {
+      assert(() {
+        _debugMutationsLocked = true;
+        return true;
+      }());
+      callback(this);
+      assert(() {
+        _debugMutationsLocked = false;
+        return true;
+      }());
+    };
+    return () {
+      _callbacks.remove(callbackId);
```

Comment:
Maybe assert that the callback is actually still in _callbacks before attempting to remove it to avoid that somebody accidentally calls the remover twice?

Reply:
Done.

Calling it twice is wasted work but it shouldn't cause any program incorrectness at least. But assert sgtm.